keysmash.c

```diff
@@ -1,17 +1,25 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <assert.h>
 
-char* keysmash(int length) {
-    char keys[] = { 'a', 's', 'd', 'f', 'j', 'k', 'l', 'a', 's', 'd', 'f', 'j', 'k', 'l', ';' };
-    length -= 2;
-    char* smash = (char*)malloc(sizeof(char)*1000);
+char* keysmash(unsigned length) {
+    char keys[16] = { 'a', 's', 'd', 'f', 'g', 'j', 'k', 'l', 'a', 's', 'd', 'f', 'j', 'k', 'l', ';' };
+    // Special case for this, otherwise we're writing out of bounds!
+    if (length == 1) {
+        return "a";
```

Comment:
This does cause issues (running with gcc because musl is still breaking things for me)

<details>
<summary>valgrind --track-origins=yes ./server</summary>

```
==28185== Memcheck, a memory error detector
==28185== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==28185== Using Valgrind-3.16.1 and LibVEX; rerun with -h for copyright info
==28185== Command: ./server
==28185== 
==28185== Invalid read of size 1
==28185==    at 0x48AB78C: ____strtol_l_internal (in /usr/lib/libc-2.33.so)
==28185==    by 0x48A8133: atoi (in /usr/lib/libc-2.33.so)
==28185==    by 0x109854: client_handler (server.c:123)
==28185==    by 0x109B65: main (server.c:305)
==28185==  Address 0x1ffefef667 is on thread 1's stack
==28185==  1961 bytes below stack pointer
==28185== 
==28185== Invalid read of size 1
==28185==    at 0x48AB7DE: ____strtol_l_internal (in /usr/lib/libc-2.33.so)
==28185==    by 0x48A8133: atoi (in /usr/lib/libc-2.33.so)
==28185==    by 0x109854: client_handler (server.c:123)
==28185==    by 0x109B65: main (server.c:305)
==28185==  Address 0x1ffefef667 is on thread 1's stack
==28185==  1961 bytes below stack pointer
==28185== 
==28185== Invalid read of size 1
==28185==    at 0x48AB865: ____strtol_l_internal (in /usr/lib/libc-2.33.so)
==28185==    by 0x48A8133: atoi (in /usr/lib/libc-2.33.so)
==28185==    by 0x109854: client_handler (server.c:123)
==28185==    by 0x109B65: main (server.c:305)
==28185==  Address 0x1ffefef668 is on thread 1's stack
==28185==  1960 bytes below stack pointer
==28185== 
==28185== Invalid read of size 1
==28185==    at 0x48AB78C: ____strtol_l_internal (in /usr/lib/libc-2.33.so)
==28185==    by 0x48A8133: atoi (in /usr/lib/libc-2.33.so)
==28185==    by 0x109867: client_handler (server.c:124)
==28185==    by 0x109B65: main (server.c:305)
==28185==  Address 0x1ffefef667 is on thread 1's stack
==28185==  1961 bytes below stack pointer
==28185== 
==28185== Invalid read of size 1
==28185==    at 0x48AB7DE: ____strtol_l_internal (in /usr/lib/libc-2.33.so)
==28185==    by 0x48A8133: atoi (in /usr/lib/libc-2.33.so)
==28185==    by 0x109867: client_handler (server.c:124)
==28185==    by 0x109B65: main (server.c:305)
==28185==  Address 0x1ffefef667 is on thread 1's stack
==28185==  1961 bytes below stack pointer
==28185== 
==28185== Invalid read of size 1
==28185==    at 0x48AB865: ____strtol_l_internal (in /usr/lib/libc-2.33.so)
==28185==    by 0x48A8133: atoi (in /usr/lib/libc-2.33.so)
==28185==    by 0x109867: client_handler (server.c:124)
==28185==    by 0x109B65: main (server.c:305)
==28185==  Address 0x1ffefef668 is on thread 1's stack
==28185==  1960 bytes below stack pointer
==28185== 
==28185== Conditional jump or move depends on uninitialised value(s)
==28185==    at 0x4841C88: strlen (vg_replace_strmem.c:459)
==28185==    by 0x1099C4: client_handler (server.c:156)
==28185==    by 0x109B65: main (server.c:305)
==28185==  Uninitialised value was created by a heap allocation
==28185==    at 0x483E77F: malloc (vg_replace_malloc.c:307)
==28185==    by 0x109897: client_handler (server.c:130)
==28185==    by 0x109B65: main (server.c:305)
==28185== 
^C==28185== 
==28185== Process terminating with default action of signal 2 (SIGINT)
==28185==    at 0x496A227: accept (in /usr/lib/libc-2.33.so)
==28185==    by 0x109B2B: main (server.c:290)
==28185== 
==28185== HEAP SUMMARY:
==28185==     in use at exit: 0 bytes in 0 blocks
==28185==   total heap usage: 3 allocs, 3 frees, 4,176 bytes allocated
==28185== 
==28185== All heap blocks were freed -- no leaks are possible
==28185== 
==28185== For lists of detected and suppressed errors, rerun with: -s
==28185== ERROR SUMMARY: 7 errors from 7 contexts (suppressed: 0 from 0)
```

</details>

Reply:
See line 136: ```c
            if (length != 1) {
                free(ks);
            }
```

Honestly, returning a char array isn't ideal either... I would like to redo this with a a pointer to a char array as a param instead.

Reply:
Well, you're returning a value that's declared on the stack, so the malloc check doesn't solve that.